File: bitmarket-common/src/main/java/com/bitwise/bitmarket/common/protocol/ExchangeRequest.java
package com.bitwise.bitmarket.common.protocol;

import com.bitwise.bitmarket.common.PeerConnection;
import com.bitwise.bitmarket.common.currency.BtcAmount;

public class ExchangeRequest {

    private final OfferId id;
    private final PeerId fromId;
    private final PeerConnection fromConnection;
    private final BtcAmount amount;

    public ExchangeRequest(
            OfferId id, PeerId fromId, PeerConnection fromConnection, BtcAmount amount) {
        this.id = id;
        this.fromId = fromId;
        this.fromConnection = fromConnection;
        this.amount = amount;
    }

    public OfferId getId() {
        return this.id;
    }

    public PeerId getFromId() {
        return this.fromId;
    }

    public PeerConnection getFromConnection() {
        return this.fromConnection;
    }

    public BtcAmount getAmount() {
        return this.amount;
    }
}

File: bitmarket-common/src/main/scala/com/bitwise/bitmarket/common/protocol/ExchangeRequest.scala
package com.bitwise.bitmarket.common.protocol

import com.bitwise.bitmarket.common.PeerConnection
import com.bitwise.bitmarket.common.currency.BtcAmount
import com.bitwise.bitmarket.common.protorpc.{Callbacks, PeerSession}
import com.bitwise.bitmarket.common.protocol.protobuf.{
  ProtobufConversions, BitmarketProtobuf => proto}

case class ExchangeRequest(
    exchangeId: String,
    fromId: PeerId,
    fromConnection: PeerConnection,
    amount: BtcAmount
)

object ExchangeRequest {

  implicit val Write = new MessageSend[ExchangeRequest] {

    def sendAsProto(msg: ExchangeRequest, session: PeerSession) = {
      val stub = proto.PeerService.newStub(session.channel)
      stub.requestExchange(
        session.controller,
        ProtobufConversions.toProtobuf(msg),
        Callbacks.noop[proto.ExchangeRequestResponse])
    }
  }
}


